roachprod: refactor Monitor

herkolategan opened the pull request:
Previously, Monitor was only able to monitor processes already started. It could
not detect a new processes if it was started after monitor started monitoring.

This change uses the new monitor scripts `monitor_local.sh` &
`monitor_remote.sh` which produces a frame of processes when any cockroach
process changes occur including when a new cockroach process is started.

The logic for detecting changes in processes has been moved to the "client"
side. The scripts are now less complicated and only have the responsibility of
sending through a frame of processes. A frame consists of the cluster label,
process id, and processes status of all cockroach processes.

In addition, Monitor has been moved into its own source file for better
separation of logic. The `ignore empty nodes` functionality has also been removed,
as it is an unreliable way of detecting "empty nodes", and does not serve any purpose.

Informs: #118214
Epic: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/136879)
<!-- Reviewable:end -->

herkolategan committed:
roachprod: use logger for monitor CLI

Previously, running roachprod Monitor, from CLI, did not include a timestamp
when an event is fired. Using a logger adds a timestamp to the output which
makes it more useful to determine when a monitor event happened.

Epic: None

# Conflicts:
#	pkg/cmd/roachprod/main.go

herkolategan committed:
roachprod: monitor local script

Add a new monitor script for monitoring local processes. On OSX development
machines `systemctl` is not available so we need to rely on process status, `ps`
command, to determine which processes are running. One limitation of using
`ps` is that we are unable to determine the status for a process that has
exited. But to keep consistent with the output Monitor expects we append a
status line "unknown" for each process.

The biggest difference between the new and old script is that this script does
not contain any logic to detect process changes, but rather only provides
information on cockroach processes back to the caller. It does however optimise
a little bit by not resending the information if it has not changed. This now
puts the responsibility on the caller (Monitor) to maintain the logic for
detecting process changes and emitting events.

Epic: None

herkolategan committed:
roachprod: monitor remote script

Add a new monitor script for monitoring remote processes. `systemctl` is used to
monitor cockroach processes, on remote nodes, and determine the exit status if a
process has died.

It emits the same frame format as the local script version, and depends on
Monitor to implement the logic for detecting changes in the process list frame.

Informs: #118214
Epic: None

herkolategan committed:
roachprod: mock session

Previously, it was difficult to test code that creates a remote session.

This change adds a mock implementation of session, and the hooks in
SyncedCluster to use it during tests.

The mock session provides several hooks for controlling the IO of a session.
The inputs and outputs are backed by a channel to allow finer control of the
flow of data.

Currently, it only implements what was required to test the Monitor
functionality, but can be extended in the future to cover more use cases.

Informs: #118214
Epic: None

herkolategan committed:
roachprod: refactor Monitor

Previously, Monitor was only able to monitor processes already started. It could
not detect a new processes if it was started after monitor started monitoring.

This change uses the new monitor scripts `monitor_local.sh` &
`monitor_remote.sh` which produces a frame of processes when any cockroach
process changes occur including when a new cockroach process is started.

The logic for detecting changes in processes has been moved to the "client"
side. The scripts are now less complicated and only have the responsibility of
sending through a frame of processes. A frame consists of the cluster label,
process id, and processes status of all cockroach processes.

In addition, Monitor has been moved into its own source file for better
separation of logic. The `ignore empty nodes` functionality has also been removed,
as it is an unreliable way of detecting "empty nodes", and does not serve any
real purpose anymore.

Informs: #118214
Epic: None

herkolategan committed:
roachtest: health status for given nodes

Previously, `cluster.HealthStatus` seemed to check all nodes, but rely on the
count of a nodes list passed in as a parameter.

This change fixes the function to use the passed nodes during the request phase
as well. It also adds a small retry mechanism during the 15 seconds it checks
the health of the nodes.

The issue was found when updating the health check to only look at
`c.CRDBNodes()` and exclude the workload node if one is part of the cluster.

Informs: #118214
Epic: None

herkolategan committed:
roachtest: remove assert dead nodes

Previously, the post assertions would check for dead nodes by using
`roachprod.Monitor`. It checks if the data directory of a cockroach node is
non-trivial. This is not a reliable way to exclude empty nodes, or to check for
dead nodes. It also complicates the Monitor functionality unnecessarily. It also
no longer functions, since `roachprod.Monitor` only watches already live
processes.

This change removes this functionality as it no longer works. An issue (See:
#137329) has been created to address this should it still be a requirement.

Informs: #118214
Epic: None

herkolategan committed:
roachprod: add Monitor data driven tests

Previously, Monitor was mostly untested. This change aims to add a few data
driven tests using the new mock session functionality. The Monitor logic for
checking for process changes have moved to the client, and can now be tested as
well.

Although we can't test that the shell scripts are correct with these tests, we
can test that Monitor interprets the output correctly, and emits the correct
events. We can also test that it correctly handles IO errors.

The data driven tests specify what the scripts output, and we can then verify
the number of events we expect and what those events are. Since events can be
out of order, to ensure the tests do not flake we sort the events. Event timing
can be tested by chaining multiple output write blocks and event checks.

Additionally, a few smalls separate tests are added to check different IO
scenarios.

Informs: #118214
Epic: None

herkolategan commented:
TeamCity Run @ 0.7: https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/18391704
There was one monitor failure, which is an actual bug (not monitor related).

herkolategan commented:
> Your pull request contains more than 1000 changes. It is strongly encouraged to split big PRs into smaller chunks.
Added a few bash scripts, and data driven tests which makes this PR look bulkier than it is.

golgeek commented on pkg/roachprod/install/monitor.go:
```diff
@@ -0,0 +1,373 @@
+	// monitor script output frame. If a process ID is not updated, it is
+	// considered dead.
+	for _, process := range m.processes {
+		process.lastProcessID = process.processID
```
Nit: maybe it could deserve a dedicated `monitorProcess.reset()` method to simplify the logic in `processMonitorOutput()`

herkolategan replied:
Done.

golgeek commented on pkg/roachprod/install/monitor.go:
```diff
@@ -0,0 +1,373 @@
+		process.status = statusLine[1]
+	}
+
+	// Check for dead processes by comparing previous process IDs.
```
Nit: maybe also move the code below to a `monitorNode.checkForDeadProcess()` method to streamline the logic in this function to only parsing the output of the `monitor_{local,remote}.sh` scripts.

herkolategan replied:
Done.

golgeek commented on pkg/roachprod/install/monitor.go:
```diff
@@ -0,0 +1,373 @@
+}
+
+func (m *monitorNode) monitorNode(ctx context.Context, l *logger.Logger) {
+	params := struct {
```
Nit: I feel like I'm only focusing on one thing (really sorry about that!), but the templating could also be extracted to its own `monitorNode.prepareTemplate()` method to streamline the logic

herkolategan replied:
I also prefer splitting these up like that. Thanks for the suggestions. Done.

golgeek reviewed (approved):
Looks great, and much cleaner and more organized than the previous code!
Only a few comments on code organization.

herkolategan commented on pkg/roachprod/install/monitor.go:
```diff
@@ -0,0 +1,373 @@
+		)
+	case MonitorError:
+		status = fmt.Sprintf("error: %s", event.Err.Error())
+	}
```
Opted for the solution provided above.

DarrylWong commented on pkg/cmd/roachtest/cluster.go:
```diff
@@ -1550,20 +1514,24 @@ func newHealthStatusResult(node int, status int, body []byte, err error) *Health
-	getStatus := func(ctx context.Context, node int) *HealthStatusResult {
-		url := fmt.Sprintf(`https://%s/health?ready=1`, adminAddrs[node-1])
+	getStatus := func(ctx context.Context, nodeIndex, node int) *HealthStatusResult {
+		url := fmt.Sprintf(`https://%s/health?ready=1`, adminAddrs[nodeIndex])
```
Looks like this was already broken before, but I think this url needs to be http if the cluster is insecure.

DarrylWong replied:
Ran `acceptance-cluster-init` which creates an insecure cluster and post test assertions fail with:

```
test-post-assertions: 2025/01/10 21:55:32 test_runner.go:1493: n1:/health?ready=1 error=Get "https://127.0.0.1:29005/health?ready=1": http: server gave HTTP response to HTTPS client
test-post-assertions: 2025/01/10 21:55:32 test_runner.go:1493: n2:/health?ready=1 error=Get "https://127.0.0.1:29003/health?ready=1": http: server gave HTTP response to HTTPS client
test-post-assertions: 2025/01/10 21:55:32 test_runner.go:1493: n3:/health?ready=1 error=Get "https://127.0.0.1:29001/health?ready=1": http: server gave HTTP response to HTTPS client
test-post-assertions: 2025/01/10 21:55:32 test_runner.go:1493: n4:/health?ready=1 error=Get "https://127.0.0.1:29007/health?ready=1": http: server gave HTTP response to HTTPS client
no live node found, skipping validation checks
```

Side note: this message that validation checks are being skipped is buried inside `test-teardown.log`. If you only looked at `test.log`, you'd never know something went wrong:

```
2025/01/10 22:09:23 test_runner.go:1354: test completed successfully
2025/01/10 22:09:23 test_runner.go:1384: running post test assertions (test-post-assertions.log)
2025/01/10 22:09:38 test_runner.go:1397: running test teardown (test-teardown.log)
2025/01/10 22:09:38 test_runner.go:1205: --- PASS: acceptance/cluster-init (64.43s)
```

Not really anything you broke so this could be done separately imo.

herkolategan replied:
Nice observation! I guess I glanced over this since this code was fixed recently. Just going to see if I can find that PR and if there was a reason not to check secure. Otherwise I'll create a new PR for this fix. Thanks!

DarrylWong reviewed (commented):
Will take another pass soon, but so far it LGTM, just flushing the few minor comments I have so far.

herkolategan commented on pkg/roachprod/install/cluster_synced.go:
```diff
@@ -605,6 +609,7 @@ func (c *SyncedCluster) Wipe(ctx context.Context, l *logger.Logger, preserveCert
 			rmCmds := []string{
 				fmt.Sprintf(`sudo find /mnt/data* -maxdepth 1 -type f -not -name %s -exec rm -f {} \;`, vm.InitializedFile),
 				`sudo rm -fr /mnt/data*/{auxiliary,local,tmp,cassandra,cockroach,cockroach-temp*,mongo-data}`,
+				`sudo rm -fr *.sh`,
```
It's on a wipe, so it should be okay. Can't remember my reason for adding this, but I don't see the harm as wipe should leave things cleaner than they were.

herkolategan replied:
On second thought, I'm going to move this to a new PR. I think the reasoning behind this was that the start-up script get re-used for restarts, but with new clusters we don't want the old scripts to be around.

herkolategan commented on pkg/roachprod/install/cluster_synced.go:
```diff
@@ -605,6 +609,7 @@ func (c *SyncedCluster) Wipe(ctx context.Context, l *logger.Logger, preserveCert
 			rmCmds := []string{
 				fmt.Sprintf(`sudo find /mnt/data* -maxdepth 1 -type f -not -name %s -exec rm -f {} \;`, vm.InitializedFile),
 				`sudo rm -fr /mnt/data*/{auxiliary,local,tmp,cassandra,cockroach,cockroach-temp*,mongo-data}`,
+				`sudo rm -fr *.sh`,
```
Good catch, should keep these two in sync.

herkolategan commented:
TFTRs!

bors r=golgeek,DarrylWong
